# Solving cases

In [1]:
import pandas as pd

In [2]:
pd.read_csv('Student_Marks.csv')

number_courses  time_study   Marks
0                3       4.508  19.202
1                4       0.096   7.734
2                4       3.133  13.811
3                6       7.909  53.018
4                8       7.811  55.299
..             ...         ...     ...
95               6       3.561  19.128
96               3       0.301   5.609
97               4       7.163  41.444
98               7       0.309  12.027
99               3       6.335  32.357

[100 rows x 3 columns]

In [4]:
pd.read_csv('ref.csv', sep=';')

ch1  ch2  ch3
0   10    1  100
1   20    2  200
2   30    3  300
3   40    4  400
4   50    5  500

In [5]:
pd.read_csv('ref.csv', delimiter=';')

ch1  ch2  ch3
0   10    1  100
1   20    2  200
2   30    3  300
3   40    4  400
4   50    5  500

In [6]:
df=pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')

In [7]:
df.shape #(jumlah baris, jumlah kolom)

(126314, 23)

In [8]:
df.shape[0] #cuma pengen tau jumlah baris, pake axis=0

126314

In [9]:
df.shape[1]

23

In [10]:
df.head() #cuma liatin 5 data teratas

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [11]:
df.head(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...   40.29483     NYK    Knicks       68   
1            0     NYK   Knicks  ...   41.70517     TRH   Huskies       66   

  opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0    1300.0  1306.7233              H            L  0.640065   NaN  
1    1300.0  1293.2767              A            W  0.359935   NaN  

[2 rows x 23 columns]

In [12]:
pd.set_option('display.max.columns', None)

In [13]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

In [14]:
df.tail() #recall 5 data terbawah

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126310         102            1     GSW   Warriors  104  1809.9791  1813.6349   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   
126313         103            1     GSW   Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309  60.309792    GSW   Warriors      103  1790.9591  1809.9791   
126310  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391   
126311  60.010067    GSW   Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW   Warriors      105  1813.6349  1822.2881   
126313  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
126309             H           L  0.546572   NaN  
126310             H           W  0.765565   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  
126313             A           W  0.518550   NaN

## Data Overview

kata-kata penjelasan sedang apa disana, ada insight apa aja yg didapat

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

In [17]:
df.describe().T #ditranspose
#statistika deskriptif

count          mean           std          min           25%  \
gameorder    126314.0  31579.000000  18231.927643     1.000000  15790.000000   
_iscopy      126314.0      0.500000      0.500002     0.000000      0.000000   
year_id      126314.0   1988.200374     17.582309  1947.000000   1975.000000   
seasongame   126314.0     43.533733     25.375178     1.000000     22.000000   
is_playoffs  126314.0      0.063857      0.244499     0.000000      0.000000   
pts          126314.0    102.729982     14.814845     0.000000     93.000000   
elo_i        126314.0   1495.236055    112.139945  1091.644500   1417.237975   
elo_n        126314.0   1495.236055    112.461687  1085.774400   1416.994900   
win_equiv    126314.0     41.707889     10.627332    10.152501     34.103035   
opp_pts      126314.0    102.729982     14.814845     0.000000     93.000000   
opp_elo_i    126314.0   1495.236055    112.139945  1091.644500   1417.237975   
opp_elo_n    126314.0   1495.236055    112.461687  1085.774400   1416.994900   
forecast     126314.0      0.500000      0.215252     0.020447      0.327989   

                      50%           75%           max  
gameorder    31579.000000  47368.000000  63157.000000  
_iscopy          0.500000      1.000000      1.000000  
year_id       1990.000000   2003.000000   2015.000000  
seasongame      43.000000     65.000000    108.000000  
is_playoffs      0.000000      0.000000      1.000000  
pts            103.000000    112.000000    186.000000  
elo_i         1500.945550   1576.060000   1853.104500  
elo_n         1500.954400   1576.291625   1853.104500  
win_equiv       42.113357     49.635328     71.112038  
opp_pts        103.000000    112.000000    186.000000  
opp_elo_i     1500.945550   1576.060000   1853.104500  
opp_elo_n     1500.954400   1576.291625   1853.104500  
forecast         0.500000      0.672011      0.979553

## Data Exploration

In [19]:
df.head(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767   40.29483    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7233   41.70517    TRH   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   
1  Huskies       66     1300.0  1293.2767             A           W  0.359935   

  notes  
0   NaN  
1   NaN

In [20]:
df['game_result'].unique()

array(['L', 'W'], dtype=object)

In [21]:
df['game_result'].value_counts()

L    63157
W    63157
Name: game_result, dtype: int64

In [25]:
df['fran_id'].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

In [30]:
df.sort_values(by='pts',ascending=False)[['year_id','fran_id','pts']]

year_id  fran_id  pts
50094     1984  Pistons  186
50095     1984  Nuggets  184
21416     1970   Pacers  177
32881     1975    Sails  176
8974      1959  Celtics  173
...        ...      ...  ...
4         1947  Falcons   33
3191      1951  Pistons   19
3190      1951   Lakers   18
26685     1973  Squires    2
26684     1973  Nuggets    0

[126314 rows x 3 columns]

In [34]:
df.loc[df['fran_id']=='Knicks', 'opp_fran'].value_counts()
#[apa yg mau di slicing, kolom apa yg ingin dibandingkan ]
#value counts buat liat berpa kali opp fran disebut untuk fran id==knicks

Celtics         526
Sixers          471
Pistons         382
Hawks           368
Warriors        335
Wizards         323
Lakers          310
Kings           299
Bulls           251
Pacers          205
Bucks           204
Cavaliers       203
Nets            189
Rockets         145
Heat            137
Thunder         124
Clippers        122
Suns            120
Trailblazers    105
Magic            99
Pelicans         90
Jazz             90
Spurs            90
Raptors          85
Nuggets          81
Mavericks        68
Baltimore        62
Timberwolves     50
Hornets          40
Grizzlies        37
Capitols         32
Stags            23
Bombers          23
Olympians        20
Steamrollers     20
Rebels            9
Falcons           6
Ironmen           6
Huskies           6
Jets              5
Packers           2
Denver            2
Redskins          2
Waterloo          2
Name: opp_fran, dtype: int64

In [38]:
df.loc[(df['year_id']==1980) & (df['fran_id']=='Knicks'),:]
#2 kondisi

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
41742      20872  197910120ATL   NBA        1     1980  10/12/1979   
41769      20885  197910130NYK   NBA        0     1980  10/13/1979   
41792      20897  197910160NYK   NBA        0     1980  10/16/1979   
41816      20909  197910180DET   NBA        1     1980  10/18/1979   
41830      20916  197910190NJN   NBA        1     1980  10/19/1979   
...          ...           ...   ...      ...      ...         ...   
43444      21723  198003220NYK   NBA        0     1980   3/22/1980   
43457      21729  198003230PHI   NBA        1     1980   3/23/1980   
43468      21735  198003250CLE   NBA        1     1980   3/25/1980   
43481      21741  198003260BOS   NBA        1     1980   3/26/1980   
43500      21751  198003270NYK   NBA        0     1980   3/27/1980   

       seasongame  is_playoffs team_id fran_id  pts      elo_i      elo_n  \
41742           1            0     NYK  Knicks  104  1417.7286  1412.8352   
41769           2            0     NYK  Knicks  125  1412.8352  1421.9812   
41792           3            0     NYK  Knicks  126  1421.9812  1428.8911   
41816           4            0     NYK  Knicks  115  1428.8911  1421.2010   
41830           5            0     NYK  Knicks  111  1421.2010  1430.3867   
...           ...          ...     ...     ...  ...        ...        ...   
43444          78            0     NYK  Knicks  113  1462.0233  1447.9253   
43457          79            0     NYK  Knicks  108  1447.9253  1445.9675   
43468          80            0     NYK  Knicks  128  1445.9675  1466.1354   
43481          81            0     NYK  Knicks  121  1466.1354  1463.9615   
43500          82            0     NYK  Knicks  101  1463.9615  1459.9078   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
41742  31.612125    ATL      Hawks      121  1559.2893  1564.1827   
41769  32.525932    WSB    Wizards      118  1536.1177  1526.9717   
41792  33.222260    HOU    Rockets      121  1517.0374  1510.1274   
41816  32.447624    DET    Pistons      129  1452.9895  1460.6796   
41830  33.373604    NJN       Nets      106  1399.9999  1390.8142   
...          ...    ...        ...      ...        ...        ...   
43444  35.163406    WSB    Wizards      122  1448.1752  1462.2732   
43457  34.962353    PHI     Sixers      112  1603.8334  1605.7911   
43468  37.045979    CLE  Cavaliers      115  1507.2769  1487.1090   
43481  36.820198    BOS    Celtics      129  1669.1906  1671.3644   
43500  36.399860    PHI     Sixers      103  1609.5419  1613.5956   

      game_location game_result  forecast notes  
41742             A           L  0.199322   NaN  
41769             H           W  0.466544   NaN  
41792             H           W  0.507114   NaN  
41816             A           L  0.328634   NaN  
41830             A           W  0.388505   NaN  
...             ...         ...       ...   ...  
43444             H           L  0.658218   NaN  
43457             A           L  0.186468   NaN  
43468             A           W  0.283215   NaN  
43481             A           L  0.148739   NaN  
43500             H           L  0.434778   NaN  

[82 rows x 23 columns]

## Fungsi Agregasi
min(), max()

In [44]:
df.loc[df['team_id']=='NYK', 'year_id'].max()

2015

In [46]:
df.loc[df['team_id']=='NYK', 'year_id'].agg(['min','max'])

min    1947
max    2015
Name: year_id, dtype: int64

team dengan fran_id Knicks bermain hanya selama tahun 1947 - 2015

In [48]:
df.sample(5) #ambil 5 data acak

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
9865        4933  196011260GSW   NBA        0     1961  11/26/1960   
73499      36750  199404200LAC   NBA        1     1994   4/20/1994   
46310      23156  198201020GSW   NBA        1     1982    1/2/1982   
62846      31424  199001130SEA   NBA        0     1990   1/13/1990   
74792      37397  199501260WSB   NBA        0     1995   1/26/1995   

       seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
9865           17            0     PHW   Warriors  138  1632.6007  1636.2716   
73499          80            0     DEN    Nuggets  100  1555.3926  1564.4733   
46310          30            0     DAL  Mavericks  100  1305.0366  1301.8625   
62846          33            0     SEA    Thunder  113  1512.3881  1518.2906   
74792          38            0     WSB    Wizards  118  1342.2546  1335.9209   

       win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
9865   53.778923    CIN     Kings      108  1366.7457  1363.0748   
73499  48.069649    LAC  Clippers       85  1358.0699  1348.9893   
46310  22.015768    GSW  Warriors      119  1539.5156  1542.6897   
62846  45.231369    ATL     Hawks      106  1524.0881  1518.1855   
74792  24.861895    GSW  Warriors      121  1407.0385  1413.3722   

      game_location game_result  forecast notes  
9865              H           W  0.891488   NaN  
73499             A           W  0.636507   NaN  
46310             A           L  0.127259   NaN  
62846             H           W  0.624407   NaN  
74792             H           L  0.550507   NaN

In [49]:
df.loc[df['fran_id']=='Warriors','pts'].sum()

591224

In [55]:
df.iloc[74521:74530,1:6] #akses dataframe berdasarkan index

game_id lg_id  _iscopy  year_id date_game
74521  199501060ORL   NBA        1     1995  1/6/1995
74522  199501070ATL   NBA        0     1995  1/7/1995
74523  199501070ATL   NBA        1     1995  1/7/1995
74524  199501070CHH   NBA        1     1995  1/7/1995
74525  199501070CHH   NBA        0     1995  1/7/1995
74526  199501070CLE   NBA        1     1995  1/7/1995
74527  199501070CLE   NBA        0     1995  1/7/1995
74528  199501070DEN   NBA        0     1995  1/7/1995
74529  199501070DEN   NBA        1     1995  1/7/1995

In [59]:
df.loc[74521:74530,'game_id': 'date_game']

game_id lg_id  _iscopy  year_id date_game
74521  199501060ORL   NBA        1     1995  1/6/1995
74522  199501070ATL   NBA        0     1995  1/7/1995
74523  199501070ATL   NBA        1     1995  1/7/1995
74524  199501070CHH   NBA        1     1995  1/7/1995
74525  199501070CHH   NBA        0     1995  1/7/1995
74526  199501070CLE   NBA        1     1995  1/7/1995
74527  199501070CLE   NBA        0     1995  1/7/1995
74528  199501070DEN   NBA        0     1995  1/7/1995
74529  199501070DEN   NBA        1     1995  1/7/1995
74530  199501070HOU   NBA        0     1995  1/7/1995

## Query Dataset

pengen ambil kondisi fran id=celtics,
                     pts>100,
                     year = 2001
                     lokasinya Home
                     hasilnya Win

In [ ]:
df[
    (df['year_id']==2001) &
    (df['fran_id']=='Celtics') &
    (df['game_location']=='H') &
    (df['game_result']=='W')
][['opp_fran','opp_pts','fran_id','pts']]

In [69]:
df.query(
    "year_id==2001 & \
    fran_id=='Celtics' & \
    game_location=='H' & \
    game_result=='W'"
    )[['opp_fran','opp_pts','fran_id','pts']]

opp_fran  opp_pts  fran_id  pts
87776    Pistons       83  Celtics  103
87805    Raptors       91  Celtics   93
87970    Wizards      109  Celtics  116
88001      Hawks       99  Celtics  106
88068    Rockets       81  Celtics   96
88093      Magic       98  Celtics  103
88233   Clippers       87  Celtics   96
88336    Nuggets      102  Celtics  104
88469       Nets       86  Celtics   99
88748       Heat       76  Celtics   88
88925      Hawks       82  Celtics   83
88975       Suns       83  Celtics   99
89013    Pistons       87  Celtics   99
89105  Cavaliers       82  Celtics  103
89141      Bucks       91  Celtics   94
89364    Thunder       82  Celtics   85
89425       Jazz       99  Celtics  103
89525    Wizards       72  Celtics  104
89756      Bulls       98  Celtics  102
90057       Nets       88  Celtics   95

## Grouping and Aggregating

In [70]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
9170         4586  195911180CIN   NBA        1     1960  11/18/1959   
26767       13384  197211020GSW   NBA        0     1973   11/2/1972   
111634      55818  200912080CHI   NBA        0     2010   12/8/2009   
23956       11979  197104100KEN   ABA        1     1971   4/10/1971   
47060       23531  198203120MIL   NBA        0     1982   3/12/1982   

        seasongame  is_playoffs team_id     fran_id  pts      elo_i  \
9170            15            0     DET     Pistons  110  1460.7488   
26767            9            0     GSW    Warriors  118  1582.5199   
111634          19            0     CHI       Bulls  101  1460.8281   
23956           89            1     FLO  Floridians  101  1395.9249   
47060           63            0     MIL       Bucks  110  1633.4910   

            elo_n  win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  \
9170    1473.9630  37.022053    CIN         Kings       93  1333.5226   
26767   1584.4091  52.792950    POR  Trailblazers      104  1293.5491   
111634  1451.8463  35.658730    NJN          Nets      103  1349.0118   
23956   1385.5282  37.547958    KEN      Colonels      118  1381.5513   
47060   1625.6387  53.955975    SEA       Thunder      112  1577.1571   

        opp_elo_n game_location game_result  forecast               notes  
9170    1320.3083             A           W  0.539102  at Indianapolis IN  
26767   1291.6599             H           W  0.903706                 NaN  
111634  1357.9937             H           L  0.771943                 NaN  
23956   1391.9480             A           L  0.379210                 NaN  
47060   1585.0094             H           L  0.710935                 NaN

In [71]:
#pengen ngegrupin yg L sama L semua yg W sama W semua
gr = df.groupby('game_result')
gr

In [72]:
gr.get_group('L')

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
6               4  194611020PRO   NBA        1     1947  11/2/1946   
8               5  194611020STB   NBA        1     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126305      63153  201506070GSW   NBA        0     2015   6/7/2015   
126306      63154  201506090CLE   NBA        1     2015   6/9/2015   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
0                1            0     TRH    Huskies   66  1300.0000  1293.2767   
3                2            0     NYK     Knicks   47  1306.7233  1297.0712   
4                1            0     DTF    Falcons   33  1300.0000  1279.6189   
6                1            0     BOS    Celtics   53  1300.0000  1294.8458   
8                1            0     PIT    Ironmen   51  1300.0000  1295.3092   
...            ...          ...     ...        ...  ...        ...        ...   
126305          99            1     GSW   Warriors   93  1806.2035  1797.5032   
126306         100            1     GSW   Warriors   91  1797.5032  1790.9591   
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   

        win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
0       40.294830    NYK        Knicks       68  1300.0000  1306.7233   
3       40.692783    CHS         Stags       63  1300.0000  1309.6521   
4       38.864048    WSC      Capitols       50  1300.0000  1320.3811   
6       40.459381    PRO  Steamrollers       59  1300.0000  1305.1542   
8       40.507980    STB       Bombers       56  1300.0000  1304.6908   
...           ...    ...           ...      ...        ...        ...   
126305  67.029549    CLE     Cavaliers       95  1708.1705  1716.8708   
126306  66.615417    CLE     Cavaliers       96  1716.8708  1723.4149   
126309  60.309792    GSW      Warriors      103  1790.9591  1809.9791   
126311  60.010067    GSW      Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW      Warriors      105  1813.6349  1822.2881   

       game_location game_result  forecast notes  
0                  H           L  0.640065   NaN  
3                  A           L  0.368899   NaN  
4                  H           L  0.640065   NaN  
6                  A           L  0.359935   NaN  
8                  A           L  0.359935   NaN  
...              ...         ...       ...   ...  
126305             H           L  0.757674   NaN  
126306             A           L  0.472157   NaN  
126309             H           L  0.546572   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  

[63157 rows x 23 columns]

In [73]:
gr.get_group('W')

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
5               3  194611020DTF   NBA        1     1947  11/2/1946   
7               4  194611020PRO   NBA        0     1947  11/2/1946   
9               5  194611020STB   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126304      63153  201506070GSW   NBA        1     2015   6/7/2015   
126307      63154  201506090CLE   NBA        0     2015   6/9/2015   
126308      63155  201506110CLE   NBA        1     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id       fran_id  pts      elo_i  \
1                1            0     NYK        Knicks   68  1300.0000   
2                1            0     CHS         Stags   63  1300.0000   
5                1            0     WSC      Capitols   50  1300.0000   
7                1            0     PRO  Steamrollers   59  1300.0000   
9                1            0     STB       Bombers   56  1300.0000   
...            ...          ...     ...           ...  ...        ...   
126304          98            1     CLE     Cavaliers   95  1708.1705   
126307          99            1     CLE     Cavaliers   96  1716.8708   
126308         101            1     GSW      Warriors  103  1790.9591   
126310         102            1     GSW      Warriors  104  1809.9791   
126313         103            1     GSW      Warriors  105  1813.6349   

            elo_n  win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
1       1306.7233  41.705170    TRH    Huskies       66  1300.0000  1293.2767   
2       1309.6521  42.012257    NYK     Knicks       47  1306.7233  1297.0712   
5       1320.3811  43.135952    DTF    Falcons       33  1300.0000  1279.6189   
7       1305.1542  41.540619    BOS    Celtics       53  1300.0000  1294.8458   
9       1304.6908  41.492020    PIT    Ironmen       51  1300.0000  1295.3092   
...           ...        ...    ...        ...      ...        ...        ...   
126304  1716.8708  61.312763    GSW   Warriors       93  1806.2035  1797.5032   
126307  1723.4149  61.826408    GSW   Warriors       91  1797.5032  1790.9591   
126308  1809.9791  67.794983    CLE  Cavaliers       82  1723.4149  1704.3949   
126310  1813.6349  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391   
126313  1822.2881  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
1                  A           W  0.359935   NaN  
2                  H           W  0.631101   NaN  
5                  A           W  0.359935   NaN  
7                  H           W  0.640065   NaN  
9                  H           W  0.640065   NaN  
...              ...         ...       ...   ...  
126304             A           W  0.242326   NaN  
126307             H           W  0.527843   NaN  
126308             A           W  0.453428   NaN  
126310             H           W  0.765565   NaN  
126313             A           W  0.518550   NaN  

[63157 rows x 23 columns]

In [76]:
df.groupby('game_result')[['pts','opp_pts']].mean()

pts     opp_pts
game_result                        
L             97.349209  108.110756
W            108.110756   97.349209

rata-rata grup yg kalah pointnya 97.3, yg menang 108.11

In [79]:
df[
    (df['year_id']>2010)&
    (df['fran_id']=='Spurs')
].groupby(['year_id','game_result'])['game_id'].count()

year_id  game_result
2011     L              25
         W              63
2012     L              20
         W              60
2013     L              30
         W              73
2014     L              27
         W              78
2015     L              31
         W              58
Name: game_id, dtype: int64

In [80]:
df[
    (df['year_id']>2010)&
    (df['fran_id']=='Spurs')
].groupby(['year_id','game_result']).agg({'pts':'mean', 'game_id': 'count'})

pts  game_id
year_id game_result                     
2011    L             93.840000       25
        W            106.698413       63
2012    L             94.350000       20
        W            106.366667       60
2013    L             93.333333       30
        W            106.232877       73
2014    L             96.925926       27
        W            108.564103       78
2015    L             97.548387       31
        W            106.224138       58

## Manipulating Columns
- create new columns based on existing data
- rename columns
- drop columns

In [81]:
#create new columns based on existing data
df['selisih poin'] = df['pts'] - df['opp_pts']

In [82]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  selisih poin  
0  0.640065   NaN            -2  
1  0.359935   NaN             2  
2  0.631101   NaN            16  
3  0.368899   NaN           -16  
4  0.640065   NaN           -17

In [83]:
nba = df.copy()

In [84]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  selisih poin  
0  0.640065   NaN            -2  
1  0.359935   NaN             2  
2  0.631101   NaN            16  
3  0.368899   NaN           -16  
4  0.640065   NaN           -17

In [86]:
#rename columns
nba.rename(columns={'game_location': 'location','game_result': 'result'}, inplace=True)
#nba_edit = nba.rename(columns={'game_location': 'location','game_result': 'result'})

In [87]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n location result  forecast notes  \
0    Knicks       68  1300.0000  1306.7233        H      L  0.640065   NaN   
1   Huskies       66  1300.0000  1293.2767        A      W  0.359935   NaN   
2    Knicks       47  1306.7233  1297.0712        H      W  0.631101   NaN   
3     Stags       63  1300.0000  1309.6521        A      L  0.368899   NaN   
4  Capitols       50  1300.0000  1320.3811        H      L  0.640065   NaN   

   selisih poin  
0            -2  
1             2  
2            16  
3           -16  
4           -17

In [90]:
#drop columns
nba.drop(['elo_i','elo_n'],axis=1, inplace = True)

In [92]:
nba['kolom kosong'] = 0
#nba['kolom kosong'] = np.NaN 
# buat bikin kolom kosong isinya Not a Number


In [93]:
nba

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts  win_equiv opp_id  \
0                1            0     TRH    Huskies   66  40.294830    NYK   
1                1            0     NYK     Knicks   68  41.705170    TRH   
2                1            0     CHS      Stags   63  42.012257    NYK   
3                2            0     NYK     Knicks   47  40.692783    CHS   
4                1            0     DTF    Falcons   33  38.864048    WSC   
...            ...          ...     ...        ...  ...        ...    ...   
126309         100            1     CLE  Cavaliers   82  60.309792    GSW   
126310         102            1     GSW   Warriors  104  68.013329    CLE   
126311         101            1     CLE  Cavaliers   91  60.010067    GSW   
126312         102            1     CLE  Cavaliers   97  59.290245    GSW   
126313         103            1     GSW   Warriors  105  68.519516    CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n location result  forecast  \
0          Knicks       68  1300.0000  1306.7233        H      L  0.640065   
1         Huskies       66  1300.0000  1293.2767        A      W  0.359935   
2          Knicks       47  1306.7233  1297.0712        H      W  0.631101   
3           Stags       63  1300.0000  1309.6521        A      L  0.368899   
4        Capitols       50  1300.0000  1320.3811        H      L  0.640065   
...           ...      ...        ...        ...      ...    ...       ...   
126309   Warriors      103  1790.9591  1809.9791        H      L  0.546572   
126310  Cavaliers       91  1704.3949  1700.7391        H      W  0.765565   
126311   Warriors      104  1809.9791  1813.6349        A      L  0.234435   
126312   Warriors      105  1813.6349  1822.2881        H      L  0.481450   
126313  Cavaliers       97  1700.7391  1692.0859        A      W  0.518550   

       notes  selisih poin  kolom kosong  
0        NaN            -2             0  
1        NaN             2             0  
2        NaN            16             0  
3        NaN           -16             0  
4        NaN           -17             0  
...      ...           ...           ...  
126309   NaN           -21             0  
126310   NaN            13             0  
126311   NaN           -13             0  
126312   NaN            -8             0  
126313   NaN             8             0  

[126314 rows x 23 columns]

In [1]:
2**2

4

In [2]:
3**3

27

In [3]:
2**3

8

In [4]:
s = 'ABCde'
s.capitalize()

'Abcde'

In [5]:
s.lower()

'abcde'

In [7]:
s.swapcase()

'abcDE'

In [9]:
s.title()

'Abcde'